File: snippets/screens/swiftui/search_screen/1/search_screen.swift
import SwiftUI
import Combine

struct SearchScreen: View {
    @State private var searchText = ""
    @State private var recentSearches: [String] = []
    @State private var suggestions: [String] = []
    @State private var isLoading = false
    @FocusState private var isSearchFocused: Bool

    private let cancellables = Set<AnyCancellable>()

    var body: some View {
        NavigationView {
            VStack {
                // Search Bar
                VStack(alignment: .leading, spacing: 16) {
                    HStack {
                        HStack {
                            Image(systemName: "magnifyingglass")
                                .foregroundColor(.gray)

                            TextField("Search for anything...", text: $searchText)
                                .focused($isSearchFocused)
                                .textFieldStyle(.plain)
                                .onSubmit {
                                    performSearch(searchText)
                                }

                            if !searchText.isEmpty {
                                Button {
                                    searchText = ""
                                    suggestions = []
                                } label: {
                                    Image(systemName: "xmark.circle.fill")
                                        .foregroundColor(.gray)
                                }
                            }
                        }
                        .padding(.horizontal, 12)
                        .padding(.vertical, 8)
                        .background(Color(.systemGray6))
                        .cornerRadius(10)
                    }

                    // Quick Filters
                    ScrollView(.horizontal, showsIndicators: false) {
                        HStack(spacing: 12) {
                            quickFilterButton("All", isSelected: true)
                            quickFilterButton("Images", isSelected: false)
                            quickFilterButton("Videos", isSelected: false)
                            quickFilterButton("Documents", isSelected: false)
                            quickFilterButton("Audio", isSelected: false)
                        }
                        .padding(.horizontal)
                    }
                }
                .padding(.horizontal)

                // Content Area
                if searchText.isEmpty {
                    recentSearchesView
                } else {
                    suggestionsView
                }

                Spacer()
            }
            .navigationTitle("Search")
            .navigationBarTitleDisplayMode(.large)
            .toolbar {
                ToolbarItem(placement: .navigationBarTrailing) {
                    NavigationLink("Advanced") {
                        AdvancedSearchScreen()
                    }
                }
            }
            .onAppear {
                loadRecentSearches()
                isSearchFocused = true
            }
            .onChange(of: searchText) { _, newValue in
                updateSuggestions(for: newValue)
            }
        }
    }

    private func quickFilterButton(_ title: String, isSelected: Bool) -> some View {
        Button(title) {
            // Handle filter selection
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 8)
        .background(isSelected ? Color.accentColor : Color(.systemGray6))
        .foregroundColor(isSelected ? .white : .primary)
        .cornerRadius(20)
    }

    private var recentSearchesView: some View {
        VStack(alignment: .leading, spacing: 0) {
            if recentSearches.isEmpty {
                Spacer()
                VStack(spacing: 16) {
                    Image(systemName: "magnifyingglass")
                        .font(.system(size: 64))
                        .foregroundColor(.gray)

                    Text("Start typing to search")
                        .font(.title2)
                        .foregroundColor(.gray)
                }
                Spacer()
            } else {
                HStack {
                    Text("Recent Searches")
                        .font(.headline)
                        .padding(.horizontal)

                    Spacer()

                    Button("Clear All") {
                        recentSearches.removeAll()
                    }
                    .padding(.horizontal)
                }
                .padding(.top)

                List {
                    ForEach(recentSearches, id: \.self) { search in
                        HStack {
                            Image(systemName: "clock")
                                .foregroundColor(.gray)

                            Text(search)

                            Spacer()

                            Button {
                                performSearch(search)
                            } label: {
                                Image(systemName: "arrow.up.right")
                                    .foregroundColor(.gray)
                            }
                        }
                        .contentShape(Rectangle())
                        .onTapGesture {
                            searchText = search
                            isSearchFocused = true
                        }
                    }
                }
                .listStyle(.plain)
            }
        }
    }

    private var suggestionsView: some View {
        VStack {
            if isLoading {
                ProgressView()
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
            } else {
                List {
                    ForEach(suggestions, id: \.self) { suggestion in
                        HStack {
                            Image(systemName: "magnifyingglass")
                                .foregroundColor(.gray)

                            Text(suggestion)

                            Spacer()

                            Button {
                                searchText = suggestion
                                isSearchFocused = true
                            } label: {
                                Image(systemName: "arrow.up.left")
                                    .foregroundColor(.gray)
                            }
                        }
                        .contentShape(Rectangle())
                        .onTapGesture {
                            performSearch(suggestion)
                        }
                    }
                }
                .listStyle(.plain)
            }
        }
    }

    private func loadRecentSearches() {
        // Load from UserDefaults in real app
        recentSearches = [
            "Swift programming",
            "iOS development",
            "SwiftUI tutorials",
            "Xcode tips",
        ]
    }

    private func updateSuggestions(for query: String) {
        guard !query.isEmpty else {
            suggestions = []
            return
        }

        isLoading = true

        // Simulate API call delay
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
            suggestions = [
                "\(query) tutorial",
                "\(query) examples",
                "\(query) best practices",
                "\(query) documentation",
            ]
            isLoading = false
        }
    }

    private func performSearch(_ query: String) {
        guard !query.trimmingCharacters(in: .whitespaces).isEmpty else { return }

        // Add to recent searches
        if let index = recentSearches.firstIndex(of: query) {
            recentSearches.remove(at: index)
        }
        recentSearches.insert(query, at: 0)
        if recentSearches.count > 10 {
            recentSearches.removeLast()
        }

        // Navigate to results screen
        // NavigationLink or programmatic navigation would go here
    }
}

struct AdvancedSearchScreen: View {
    var body: some View {
        Text("Advanced Search")
            .navigationTitle("Advanced Search")
    }
}

#Preview {
    SearchScreen()
}

File: snippets/screens/flutter/search_screen/3/search_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class SearchScreen extends StatefulWidget {
  const SearchScreen({super.key});

  @override
  State<SearchScreen> createState() => _SearchScreenState();
}

class _SearchScreenState extends State<SearchScreen> {
  final TextEditingController _searchController = TextEditingController();
  final FocusNode _searchFocusNode = FocusNode();
  List<String> _recentSearches = [];
  List<String> _suggestions = [];
  bool _isLoading = false;

  @override
  void initState() {
    super.initState();
    _loadRecentSearches();
    _searchController.addListener(_onSearchChanged);
  }

  @override
  void dispose() {
    _searchController.dispose();
    _searchFocusNode.dispose();
    super.dispose();
  }

  void _loadRecentSearches() {
    // Load from SharedPreferences in real app
    setState(() {
      _recentSearches = [
        'Flutter tutorials',
        'Dart programming',
        'Mobile development',
        'UI design patterns',
      ];
    });
  }

  void _onSearchChanged() {
    final query = _searchController.text;
    if (query.isEmpty) {
      setState(() => _suggestions = []);
      return;
    }

    // Simulate API call for suggestions
    setState(() => _isLoading = true);
    Future.delayed(const Duration(milliseconds: 300), () {
      if (mounted) {
        setState(() {
          _suggestions = [
            '$query tutorial',
            '$query examples',
            '$query best practices',
            '$query documentation',
          ];
          _isLoading = false;
        });
      }
    });
  }

  void _performSearch(String query) {
    if (query.trim().isEmpty) return;

    // Add to recent searches
    setState(() {
      _recentSearches.remove(query);
      _recentSearches.insert(0, query);
      if (_recentSearches.length > 10) {
        _recentSearches.removeLast();
      }
    });

    // Navigate to results screen
    Navigator.pushNamed(
      context,
      '/search-results',
      arguments: {'query': query},
    );
  }

  void _clearSearch() {
    _searchController.clear();
    setState(() => _suggestions = []);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Search'),
        elevation: 0,
        backgroundColor: Theme.of(context).scaffoldBackgroundColor,
        foregroundColor: Theme.of(context).textTheme.titleLarge?.color,
        actions: [
          IconButton(
            icon: const Icon(Icons.tune),
            onPressed: () => Navigator.pushNamed(context, '/advanced-search'),
            tooltip: 'Advanced Search',
          ),
        ],
      ),
      body: Column(
        children: [
          // Search Bar
          Padding(
            padding: const EdgeInsets.all(16.0),
            child: TextField(
              controller: _searchController,
              focusNode: _searchFocusNode,
              autofocus: true,
              decoration: InputDecoration(
                hintText: 'Search for anything...',
                prefixIcon: const Icon(Icons.search),
                suffixIcon: _searchController.text.isNotEmpty
                    ? IconButton(
                        icon: const Icon(Icons.clear),
                        onPressed: _clearSearch,
                      )
                    : null,
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(12),
                ),
                filled: true,
                fillColor: Theme.of(context).colorScheme.surface,
              ),
              textInputAction: TextInputAction.search,
              onSubmitted: _performSearch,
            ),
          ),

          // Quick Filters
          Container(
            height: 40,
            margin: const EdgeInsets.symmetric(horizontal: 16),
            child: ListView(
              scrollDirection: Axis.horizontal,
              children: [
                _buildQuickFilter('All', true),
                _buildQuickFilter('Images', false),
                _buildQuickFilter('Videos', false),
                _buildQuickFilter('Documents', false),
                _buildQuickFilter('Audio', false),
              ],
            ),
          ),

          const SizedBox(height: 16),

          // Content Area
          Expanded(
            child: _searchController.text.isEmpty
                ? _buildRecentSearches()
                : _buildSuggestions(),
          ),
        ],
      ),
    );
  }

  Widget _buildQuickFilter(String label, bool isSelected) {
    return Container(
      margin: const EdgeInsets.only(right: 8),
      child: FilterChip(
        label: Text(label),
        selected: isSelected,
        onSelected: (selected) {
          // Handle filter selection
        },
        backgroundColor: Theme.of(context).colorScheme.surface,
        selectedColor: Theme.of(context).colorScheme.primaryContainer,
      ),
    );
  }

  Widget _buildRecentSearches() {
    if (_recentSearches.isEmpty) {
      return const Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Icon(Icons.search, size: 64, color: Colors.grey),
            SizedBox(height: 16),
            Text(
              'Start typing to search',
              style: TextStyle(fontSize: 18, color: Colors.grey),
            ),
          ],
        ),
      );
    }

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Padding(
          padding: const EdgeInsets.symmetric(horizontal: 16),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              Text(
                'Recent Searches',
                style: Theme.of(context).textTheme.titleMedium,
              ),
              TextButton(
                onPressed: () {
                  setState(() => _recentSearches.clear());
                },
                child: const Text('Clear All'),
              ),
            ],
          ),
        ),
        Expanded(
          child: ListView.builder(
            itemCount: _recentSearches.length,
            itemBuilder: (context, index) {
              final search = _recentSearches[index];
              return ListTile(
                leading: const Icon(Icons.history),
                title: Text(search),
                trailing: IconButton(
                  icon: const Icon(Icons.arrow_outward),
                  onPressed: () => _performSearch(search),
                ),
                onTap: () {
                  _searchController.text = search;
                  _searchFocusNode.requestFocus();
                },
              );
            },
          ),
        ),
      ],
    );
  }

  Widget _buildSuggestions() {
    if (_isLoading) {
      return const Center(child: CircularProgressIndicator());
    }

    return ListView.builder(
      itemCount: _suggestions.length,
      itemBuilder: (context, index) {
        final suggestion = _suggestions[index];
        return ListTile(
          leading: const Icon(Icons.search),
          title: Text(suggestion),
          trailing: IconButton(
            icon: const Icon(Icons.north_west),
            onPressed: () {
              _searchController.text = suggestion;
              _searchFocusNode.requestFocus();
            },
          ),
          onTap: () => _performSearch(suggestion),
        );
      },
    );
  }
}